File: sdk/js/src/modules/nosql.ts
// sdk/js/src/modules/nosql.ts

import { BaseFetch } from "../utils/fetch";
import type {
  NoSQLFilter,
  NoSQLUpdate,
  NoSQLSort,
} from "../types/filters";
import type {
  Document,
  InsertManyResult,
  DeleteResult,
  ResponseMeta,
} from "../types/index";

// ─── Result types ─────────────────────────────────────────────────────────────

export interface FindResult<T extends Document = Document> {
  data: T[];
  meta: ResponseMeta;
}

// ─── Find Builder ─────────────────────────────────────────────────────────────

export class FindBuilder<T extends Document = Document> {
  private http: BaseFetch;
  private projectId: string;
  private collectionName: string;

  private _filter: NoSQLFilter = {};
  private _sort: NoSQLSort | null = null;
  private _limit: number = 100;
  private _skip: number = 0;

  constructor(http: BaseFetch, projectId: string, collection: string) {
    this.http = http;
    this.projectId = projectId;
    this.collectionName = collection;
  }

  // ─── Chainable config ──────────────────────────────────────────────────────

  /**
   * MongoDB-style filter document.
   * @example .filter({ category: "tech", published: true })
   * @example .filter({ age: { $gte: 18 } })
   */
  filter(filter: NoSQLFilter): this {
    this._filter = filter;
    return this;
  }

  /**
   * Sort by one or more fields.
   * Use 1 for ascending, -1 for descending.
   * @example .sort({ createdAt: -1 })
   */
  sort(sort: NoSQLSort): this {
    this._sort = sort;
    return this;
  }

  /**
   * Maximum number of documents to return (1–1000, default 100).
   */
  limit(n: number): this {
    this._limit = n;
    return this;
  }

  /**
   * Number of documents to skip (for pagination).
   */
  skip(n: number): this {
    this._skip = n;
    return this;
  }

  // ─── Terminal: execute find ────────────────────────────────────────────────

  /**
   * Execute the find query and return matching documents.
   *
   * @example
   * const { data, meta } = await baas.nosql("articles")
   *   .find({ category: "tech" })
   *   .sort({ createdAt: -1 })
   *   .limit(10)
   *   .execute()
   */
  async execute(): Promise<FindResult<T>> {
    const query: Record<string, string | number | boolean | null | undefined> =
      {
        limit: this._limit,
        skip: this._skip,
      };

    // Encode sort as sort_field + sort_dir query params
    // Backend accepts: ?sort_field=createdAt&sort_dir=-1
    if (this._sort) {
      const entries = Object.entries(this._sort);
      if (entries.length > 0) {
        const [field, dir] = entries[0];
        query["sort_field"] = field;
        query["sort_dir"] = dir;
      }
    }

    // Filter is sent as a JSON body via POST to the find endpoint.
    // The backend's GET endpoint doesn't accept a filter body, so we
    // use the filter query param for simple cases and fall back to POST
    // aggregate for complex filters. For the SDK we POST to the collection
    // route with the filter encoded as query param JSON.
    // Actually the backend GET /nosql/{projectId}/collections/{collection}
    // does NOT accept a filter body — it only supports sort/limit/skip.
    // For filter support we use the aggregate pipeline endpoint.
    if (Object.keys(this._filter).length > 0) {
      return this._executeViaAggregate();
    }

    const res = await this.http.get<T[]>(
      `/v1/nosql/${this.projectId}/collections/${this.collectionName}`,
      query
    );

    return {
      data: res.data,
      meta: res.meta ?? {},
    };
  }

  /** Fall back to aggregation pipeline when a filter is provided */
  private async _executeViaAggregate(): Promise<FindResult<T>> {
    const pipeline: NoSQLFilter[] = [{ $match: this._filter }];

    if (this._sort) {
      pipeline.push({ $sort: this._sort });
    }

    if (this._skip > 0) {
      pipeline.push({ $skip: this._skip });
    }

    pipeline.push({ $limit: this._limit });

    const res = await this.http.post<T[]>(
      `/v1/nosql/${this.projectId}/collections/${this.collectionName}/aggregate`,
      { pipeline }
    );

    return {
      data: res.data,
      meta: { count: res.data.length, limit: this._limit, skip: this._skip },
    };
  }
}

// ─── CollectionBuilder ────────────────────────────────────────────────────────

/**
 * Entry point for all operations on a single collection.
 * Returned by `baas.nosql("collection-name")`.
 */
export class CollectionBuilder<T extends Document = Document> {
  private http: BaseFetch;
  private projectId: string;
  private collectionName: string;

  constructor(http: BaseFetch, projectId: string, collection: string) {
    this.http = http;
    this.projectId = projectId;
    this.collectionName = collection;
  }

  // ─── Find ──────────────────────────────────────────────────────────────────

  /**
   * Start a find query with an optional filter.
   *
   * @example
   * const { data } = await baas.nosql("articles")
   *   .find({ category: "tech" })
   *   .sort({ createdAt: -1 })
   *   .limit(10)
   *   .execute()
   */
  find(filter: NoSQLFilter = {}): FindBuilder<T> {
    return new FindBuilder<T>(
      this.http,
      this.projectId,
      this.collectionName
    ).filter(filter);
  }

  /**
   * Fetch a single document by its id.
   * Returns `null` if not found.
   */
  async findById(id: string): Promise<T | null> {
    try {
      const res = await this.http.get<T>(
        `/v1/nosql/${this.projectId}/collections/${this.collectionName}/${id}`
      );
      return res.data;
    } catch (err: unknown) {
      if (
        err instanceof Error &&
        "statusCode" in err &&
        (err as { statusCode?: number }).statusCode === 404
      ) {
        return null;
      }
      throw err;
    }
  }

  // ─── Insert ────────────────────────────────────────────────────────────────

  /**
   * Insert a single document.
   * @example const doc = await baas.nosql("articles").insertOne({ title: "Hello" })
   */
  async insertOne(data: Omit<T, "id">): Promise<T> {
    const res = await this.http.post<T>(
      `/v1/nosql/${this.projectId}/collections/${this.collectionName}`,
      { data }
    );
    return res.data;
  }

  /**
   * Insert multiple documents in a single request.
   * @example const result = await baas.nosql("articles").insertMany([{ title: "A" }, { title: "B" }])
   */
  async insertMany(
    docs: Array<Omit<T, "id">>
  ): Promise<InsertManyResult> {
    const res = await this.http.post<{ inserted_ids: string[]; count?: number }>(
      `/v1/nosql/${this.projectId}/collections/${this.collectionName}`,
      { data: docs }
    );

    return {
      insertedIds: res.data.inserted_ids,
      count: res.data.inserted_ids.length,
    };
  }

  // ─── Update ────────────────────────────────────────────────────────────────

  /**
   * Update a document by id using MongoDB update operators.
   * If no operator is provided, wraps the update in `$set` automatically.
   *
   * @example await baas.nosql("articles").updateOne("doc-id", { $set: { title: "Updated" } })
   * @example await baas.nosql("articles").updateOne("doc-id", { title: "Updated" }) // auto $set
   */
  async updateOne(id: string, update: NoSQLUpdate): Promise<T> {
    // Auto-wrap plain objects in $set if no operator present
    const hasOperator = Object.keys(update).some((k) => k.startsWith("$"));
    const payload = hasOperator ? update : { $set: update };

    const res = await this.http.patch<T>(
      `/v1/nosql/${this.projectId}/collections/${this.collectionName}/${id}`,
      { update: payload }
    );
    return res.data;
  }

  // ─── Delete ────────────────────────────────────────────────────────────────

  /**
   * Delete a document by id.
   */
  async deleteOne(id: string): Promise<DeleteResult> {
    const res = await this.http.delete<{ deleted: boolean; id: string }>(
      `/v1/nosql/${this.projectId}/collections/${this.collectionName}/${id}`
    );
    return res.data;
  }

  // ─── Aggregation ──────────────────────────────────────────────────────────

  /**
   * Run a MongoDB aggregation pipeline.
   * @example
   * const results = await baas.nosql("orders").aggregate([
   *   { $match: { status: "paid" } },
   *   { $group: { _id: "$userId", total: { $sum: "$amount" } } },
   *   { $sort: { total: -1 } },
   * ])
   */
  async aggregate(pipeline: NoSQLFilter[]): Promise<T[]> {
    const res = await this.http.post<T[]>(
      `/v1/nosql/${this.projectId}/collections/${this.collectionName}/aggregate`,
      { pipeline }
    );
    return res.data;
  }
}

// ─── NoSQLModule ──────────────────────────────────────────────────────────────

export class NoSQLModule {
  private http: BaseFetch;
  private projectId: string;

  constructor(http: BaseFetch, projectId: string) {
    this.http = http;
    this.projectId = projectId;
  }

  /**
   * Returns a CollectionBuilder for the given collection name.
   *
   * @example
   * baas.nosql("articles").find({ published: true }).limit(20).execute()
   * baas.nosql("articles").insertOne({ title: "Hi" })
   * baas.nosql("articles").updateOne("id", { $set: { title: "Updated" } })
   * baas.nosql("articles").deleteOne("id")
   * baas.nosql("articles").aggregate([{ $match: { status: "paid" } }])
   */
  collection<T extends Document = Document>(
    name: string
  ): CollectionBuilder<T> {
    return new CollectionBuilder<T>(this.http, this.projectId, name);
  }
}

File: sdk/python/src/baas/modules/nosql.py
# sdk/python/src/yourbaas/modules/nosql.py
from __future__ import annotations

from typing import Any

from baas.types import NoSQLListResult
from baas.utils.http import HttpClient


class FindBuilder:
    """
    Chainable find query builder for NoSQL collections.

    Usage:
        result = await baas.nosql("articles") \\
            .find({"category": "tech"}) \\
            .sort("created_at", -1) \\
            .limit(10) \\
            .execute()
    """

    def __init__(self, http: HttpClient, project_id: str, collection: str) -> None:
        self._http = http
        self._project_id = project_id
        self._collection = collection
        self._filter: dict[str, Any] | None = None
        self._sort_field: str | None = None
        self._sort_dir: int = -1
        self._limit_val: int = 100
        self._skip_val: int = 0

    def _base(self) -> str:
        return f"/v1/nosql/{self._project_id}/collections/{self._collection}"

    def find(self, filter_doc: dict[str, Any] | None = None) -> "FindBuilder":
        """Set a MongoDB-style filter document."""
        self._filter = filter_doc
        return self

    def sort(self, field: str, direction: int = -1) -> "FindBuilder":
        """Set sort field and direction. -1 = descending, 1 = ascending."""
        self._sort_field = field
        self._sort_dir = direction
        return self

    def limit(self, n: int) -> "FindBuilder":
        """Limit the number of documents returned."""
        self._limit_val = n
        return self

    def skip(self, n: int) -> "FindBuilder":
        """Skip the first n documents."""
        self._skip_val = n
        return self

    async def execute(self) -> NoSQLListResult:
        """Execute the find query."""
        params: dict[str, Any] = {
            "limit": self._limit_val,
            "skip": self._skip_val,
        }
        if self._sort_field:
            params["sort_field"] = self._sort_field
            params["sort_dir"] = self._sort_dir

        resp = await self._http._get_client().get(
            self._base(),
            params=params,
            headers=self._http._build_headers(),
        )
        body = resp.json() if resp.status_code < 400 else {}
        data = body.get("data", [])
        meta = body.get("meta", {})
        return NoSQLListResult(
            data=data if isinstance(data, list) else [],
            count=meta.get("count", len(data) if isinstance(data, list) else 0),
            limit=meta.get("limit", self._limit_val),
            skip=meta.get("skip", self._skip_val),
        )


class CollectionBuilder:
    """
    Full CRUD interface for a single MongoDB collection.

    Usage:
        # Chainable find
        result = await baas.nosql("articles").find({"status": "published"}).limit(10).execute()

        # Single document
        doc = await baas.nosql("articles").get("doc-id")

        # Insert
        doc = await baas.nosql("articles").insert_one({"title": "Hello"})
        ids = await baas.nosql("articles").insert_many([{"title": "A"}, {"title": "B"}])

        # Update
        doc = await baas.nosql("articles").update_one("doc-id", {"$set": {"title": "Updated"}})

        # Delete
        await baas.nosql("articles").delete_one("doc-id")

        # Aggregate
        results = await baas.nosql("articles").aggregate([{"$match": {...}}, {"$group": {...}}])
    """

    def __init__(self, http: HttpClient, project_id: str, collection: str) -> None:
        self._http = http
        self._project_id = project_id
        self._collection = collection

    def _base(self) -> str:
        return f"/v1/nosql/{self._project_id}/collections/{self._collection}"

    # ── Find (chainable) ──────────────────────────────────────────────────────

    def find(self, filter_doc: dict[str, Any] | None = None) -> FindBuilder:
        """Return a FindBuilder for chaining sort/limit/skip/execute."""
        builder = FindBuilder(self._http, self._project_id, self._collection)
        if filter_doc:
            builder.find(filter_doc)
        return builder

    # ── CRUD ─────────────────────────────────────────────────────────────────

    async def get(self, doc_id: str) -> dict[str, Any]:
        """Fetch a single document by ID."""
        return await self._http.get(f"{self._base()}/{doc_id}")

    async def insert_one(self, data: dict[str, Any]) -> dict[str, Any]:
        """Insert a single document. Returns the created document."""
        return await self._http.post(self._base(), json_body={"data": data})

    async def insert_many(self, docs: list[dict[str, Any]]) -> list[str]:
        """Insert multiple documents. Returns a list of inserted IDs."""
        result = await self._http.post(self._base(), json_body={"data": docs})
        if isinstance(result, dict):
            return result.get("inserted_ids", [])
        return []

    async def update_one(self, doc_id: str, update: dict[str, Any]) -> dict[str, Any]:
        """
        Update a document by ID.
        update should use MongoDB update operators, e.g. {"$set": {"title": "New"}}.
        """
        return await self._http.patch(f"{self._base()}/{doc_id}", json_body={"update": update})

    async def delete_one(self, doc_id: str) -> dict[str, Any]:
        """Delete a document by ID."""
        return await self._http.delete(f"{self._base()}/{doc_id}")

    async def aggregate(self, pipeline: list[dict[str, Any]]) -> list[dict[str, Any]]:
        """Run a MongoDB aggregation pipeline."""
        result = await self._http.post(
            f"{self._base()}/aggregate",
            json_body={"pipeline": pipeline},
        )
        return result if isinstance(result, list) else []


class NoSQLModule:
    """
    Entry point for NoSQL (MongoDB) operations.

    baas.nosql("collection_name") → CollectionBuilder
    """

    def __init__(self, http: HttpClient, project_id: str) -> None:
        self._http = http
        self._project_id = project_id

    def __call__(self, collection: str) -> CollectionBuilder:
        """Return a CollectionBuilder for the given collection."""
        return CollectionBuilder(self._http, self._project_id, collection)
